boards/hail/src/main.rs

```diff
@@ -393,6 +390,13 @@ pub unsafe fn reset_handler() {
     // );
     // sam4l::gpio::PA[16].set_client(debug_process_restart);
 
+    // Configure application fault policy
+    let restart_policy = static_init!(
+        kernel::procs::ThresholdRestartThenPanic,
+        kernel::procs::ThresholdRestartThenPanic::new(4)
+    );
```

Comment:
I'm wondering why we need to statically allocate a trivial struct containing a `usize`. Can't this be just a value on the stack?

Reply:
I just default to static_init in main.rs.

Reply:
Now that I think about it, this stems from the fact that `FaultResponse` takes a `&dyn` reference, which then has to take a `'static` lifetime (because everything in the kernel assumes a `'static` lifetime), which therefore has to be created with `static_init`.

In general, I'm not sure using dynamic trait objects everywhere is the optimal design decision (we can use trait *parameters* instead, which avoids dynamic dispatch and unused code), but this seems like a much broader discussion (in this example, there's a fixed fault response throughout the lifetime of the program, as defined in each board, so dynamic dispatch could be avoided altogether in fault responses).

Reply:
@gendx indeed a broader discussion. In general, the tradeoff is between "performance" (also code size and memory) vs code readability. And some point we chose code readability and didn't really measure the impact.

It's not unlikely there are elegant ways of getting (almost) the best of both worlds, and it's also unclear how big of an impact this has on "performance". Worth exploring IMO.

Reply:
Yes, would be worth adding to the discussion in #1663.